feat: Add path option to specify additional CUE files

stuartcarnie opened the pull request:
This PR adds the ability to specify additional CUE files to unify with the base configuration. An example use case it refer to additional options in an alternate directory that will be unified with a base configuration. For example, `gopm` is invoked with the `-c` argument referring to a directory containing the following `.cue` file:

```cue
opt: path: *"" | string
```

Additionally, the `-p` argument may refer to an additional file, such as:

```cue
opt: path: "/to/some/path"
```

Resulting in a unified version on the configuration `opt: path: "/to/some/path"`

This feature will be useful for separating configuration that is specific to a user / system from a base set of configuration files.

Signed-off-by: Stuart Carnie <stuart.carnie@gmail.com>

stuartcarnie committed:
feat: Add path option to specify additional CUE files

Signed-off-by: Stuart Carnie <stuart.carnie@gmail.com>

rogpeppe commented on cmd/gopm/main.go:
```diff
@@ -108,6 +110,7 @@ func init() {
 func Main() int {
 	rootCmd.PersistentFlags().StringVarP(&rootOpt.Configuration, "config", "c", "", "Configuration directory")
 	rootCmd.PersistentFlags().StringArrayVarP(&rootOpt.Tags, "inject", "t", nil, "Set the value of a tagged field in the configuration (for example -t someField=someValue)")
+	rootCmd.PersistentFlags().StringArrayVarP(&rootOpt.CueFiles, "cue", "p", nil, "Paths to load additional CUE files")
```
Maybe say how to specify multiple files? (comma-separated? multiple `-p` options?)

stuartcarnie replied:
Done

rogpeppe commented on config/config.go:
```diff
@@ -77,19 +78,42 @@ func load0(configDir string, tags []string) (*Config, error) {
+			return nil, fmt.Errorf("CUE path must be a file, found directory %q", path)
+		}
+
+		by, err := os.ReadFile(path)
```
I think I'd probably just do this and not worry about all the explicit error message customisation above.

stuartcarnie replied:
Done 👍🏻

rogpeppe commented on config/config.go:
```diff
@@ -77,19 +78,42 @@ func load0(configDir string, tags []string) (*Config, error) {
+
+		by, err := os.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("cannot read CUE path %q: %w", path, err)
```
I'm pretty sure the error will already have the path in, so this will stutter.

```suggestion
			return nil, fmt.Errorf("cannot read CUE file: %w", path, err)
```

rogpeppe commented on cmd/gopm/testdata/gopm-path-not-exists.txt:
```diff
@@ -0,0 +1,19 @@
+}
+
+-- gopmextra/options.cue --
+package config
```
I'd be interested to know whether this package directive is needed. Kinda related: I'm wondering whether it might be nice to accept non-CUE files too e.g. JSON) so that it's easy to integrate with other kinds of data source.

stuartcarnie replied:
It is not required 👏🏻

rogpeppe reviewed (approved):
Thanks for doing this - it's been on my TODO list for a while, but I couldn't quite work out exactly which direction to go. LGTM with a few minor comments/suggestions,

stuartcarnie committed:
Update config/config.go

Co-authored-by: Roger Peppe <rogpeppe@gmail.com>

stuartcarnie committed:
chore: Cleanup error handling; update arg name

The change to argument name from cue to import is to clarify that
additional file types may be supported in the future.

Signed-off-by: Stuart Carnie <stuart.carnie@gmail.com>

stuartcarnie committed:
Merge remote-tracking branch 'origin/sgc/paths' into sgc/paths

# Conflicts:
#	config/config.go

stuartcarnie committed:
chore: Update tests with upstream error message change

Signed-off-by: Stuart Carnie <stuart.carnie@gmail.com>
